src/vos/vos_io.c

```diff
@@ -2385,6 +2384,38 @@ update_cancel(struct vos_io_context *ioc)
 			  true /* abort */);
 }
 
+/**
+ * Insert object ID and its parent container into the array of objects touched by the ongoing
+ * local transaction.
+ */
+static int
+vos_insert_oid(struct dtx_handle *dth, struct vos_container *cont, daos_unit_oid_t *oid)
+{
+	/** The array has to grow to accommodate the next record. */
+	if (dth->dth_local_oid_cnt == dth->dth_local_oid_cap) {
+		struct dtx_local_oid_record *oid_array;
+
+		D_ALLOC_ARRAY(oid_array, dth->dth_local_oid_cap << 1);
```

Comment:
any reason not to use D_REALLOC_ARRAY?   This will do the memcpy for you so may be less error prone